content/blog/unsafe-string-interning-in-go.md

```diff
@@ -0,0 +1,345 @@
+	}
+	if v != nil {
+		return v
+	}
+
+	// Value must be created now and then interned. We store a uintptr to
+	// maintain a weak reference so that the value can eventually be garbage
+	// collected if it isn't used for a period of time.
+	v = &Value{cmpVal: cmpVal}
+	val[cmpVal] = uintptr(unsafe.Pointer(v))
+	runtime.SetFinalizer(v, finalize)
+
+	return v
+}
+
+// finalize is a Go finalizer for *Value.
+func finalize(v *Value) {
+	mu.Lock()
+	defer mu.Unlock()
+	if v.resurrected {
+		// We lost the race. Somebody resurrected it while we were about to
+		// finalize it. Try again next round.
+		v.resurrected = false
+		runtime.SetFinalizer(v, finalize)
+		return
+	}
+	delete(val, v.cmpVal)
+}
+```
+
+The code comments above discuss many of the ideas at work, but let's break
+down `Get` and `finalize` specifically to discuss how these interact with each
+other and the Go runtime.
+
+Starting with `Get`, we acquire the mutex around the map of interned values,
+which now stores the values as [weak references using
+`uintptr`](https://en.wikipedia.org/wiki/Weak_reference) rather than a concrete
+`*Value` type. This ensures that the values are not protected from garbage
+collection, and thus we must manually manage their lifecycles by using
+finalizers and the `v.resurrected` boolean field.
+
+First we check if the value is already present in the pool. If it is, we convert
+the weak reference back to a `*Value` by using `unsafe.Pointer` conversion, and
+mark the value as resurrected so that any finalizers which run concurrently will
+know that this value cannot be removed from the pool.
+
+```go
+	var v *Value
+	addr, ok := val[cmpVal]; ok {
+		// Value is already interned in the pool. In case a finalizer runs in
+		// the near future, we must "resurrect" this value to bring it back
+		// to the active part of its lifecycle.
+		v = (*Value)((unsafe.Pointer)(addr))
+		v.resurrected = true
+	}
+	if v != nil {
+		return v
+	}
+```
+
+If the value did not already exist in the pool, we can create it now and intern
+it there by adding it to the map of weak references. Note that the
+`v.resurrected` field defaults to `false`, meaning nobody has bothered to
+retrieve this value from the pool as of yet. We also install the finalizer for
+`*Value`.
+
+```go
+	// Value must be created now and then interned. We store a uintptr to
+	// maintain a weak reference so that the value can eventually be garbage
+	// collected if it isn't used for a period of time.
+	v = &Value{cmpVal: cmpVal}
+	val[cmpVal] = uintptr(unsafe.Pointer(v))
+	runtime.SetFinalizer(v, finalize)
+```
+
+Finalizers in Go come with [many
+caveats](https://golang.org/pkg/runtime/#SetFinalizer) but we make use of them
+here in order to ensure that weak references in the map can be deleted when they
+are no longer needed, and thus the underlying `*Value` can be garbage collected
+from the heap at some point in the future.
+
+To avoid a race, we must verify the `resurrected` field and set another
```

Comment:
I suggest expanding this "to avoid a race", and walk people through what precisely the race is. I think I outlined the steps that need to happen for the "bad" version of the code to fail in one of the bugs on github. When I first encountered this package, it wasn't clear to me where the race was, or why `resurrected` avoids it. I expect many readers are going to be in the same situation :)